pkg/vif/endpoint.go

```diff
@@ -0,0 +1,117 @@
+package vif
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"gvisor.dev/gvisor/pkg/bufferv2"
+	"gvisor.dev/gvisor/pkg/tcpip"
+	"gvisor.dev/gvisor/pkg/tcpip/header"
+	"gvisor.dev/gvisor/pkg/tcpip/link/channel"
+	"gvisor.dev/gvisor/pkg/tcpip/stack"
+
+	"github.com/datawire/dlib/dlog"
+	"github.com/telepresenceio/telepresence/v2/pkg/vif/buffer"
+)
+
+type endpoint struct {
+	*channel.Endpoint
+	ctx     context.Context
+	wg      sync.WaitGroup
+	dev     *Device
+	closing *int32
+}
+
+const defaultDevMtu = 1500
+
+// Queue length for outbound packet, arriving at fd side for read. Overflow
+// causes packet drops. gVisor implementation-specific.
+const defaultDevOutQueueLen = 1024
+
+var _ stack.LinkEndpoint = (*endpoint)(nil)
+
+func NewEndpoint(ctx context.Context, dev *Device, closing *int32) stack.LinkEndpoint {
+	return &endpoint{
+		Endpoint: channel.New(defaultDevOutQueueLen, defaultDevMtu, ""),
+		ctx:      ctx,
+		dev:      dev,
+		closing:  closing,
+	}
+}
+
+func (d *endpoint) Attach(dp stack.NetworkDispatcher) {
+	d.Endpoint.Attach(dp)
+	if dp == nil {
+		// Stack is closing
+		return
+	}
+	d.wg.Add(2)
+	dlog.Info(d.ctx, "Starting Endpoint")
+	ctx, cancel := context.WithCancel(d.ctx)
+	go d.tunToDispatch(cancel)
+	go d.dispatchToTun(ctx)
+}
+
+func (d *endpoint) Wait() {
+	d.wg.Wait()
+	dlog.Info(d.ctx, "Endpoint done")
```

Comment:
should we/can we do Wait on d.Endpoint as well?

Reply:
There's no point. It looks like this:
```
func (*Endpoint) Wait() {}
```
The intention is that other endpoints can implement their own wait, which is what I've done.